### ISSUE TITLE
Gpu memory problem. In training phase, it will gradually occupy all the gpu memory.

### ISSUE THREAD

**Issue by weitaoatvison** [tamakoji/pytorch-syncbn#1]
@tamakoji Hi，I have tested your code on my pc. And I found when start training, the program will gradually occupy all the gpu memory and crash at last. I have tested on the code "test.py" you provided  and also tested on my code. So, could you give me some tips?
Env:
cuda 9.0 
pytorch 0.4.0
cudnn is automatically installed by pytorch 0.4.0
system: centos 7.2.1511

**Comment by tamakoji** [comment]
@weitaoatvison I have applied to quite few models (ResNet,ResNeXt,DenseNet,SE,etc..) but works fine. for test.py it should only consume at around 600MB max and won't increase gradually. what base architecture(backbone) are you using?
BTW I only tested with CUDA9.1
```
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|    0     21752      C   python3                                      585MiB |
|    1     21752      C   python3                                      585MiB |
|    2     21752      C   python3                                      585MiB |
+-----------------------------------------------------------------------------+
```

**Comment by weitaoatvison** [comment]
@tamakoji I fisrt tested on ResNet101, But it crashed at last. So I checked the test.py, and it also crashed. I will try to test it on cuda9.1 later.

**Comment by weitaoatvison** [comment]
@tamakoji Hi, I have tested the code on cuda9.1, the problem still exist. Now, the env is:
cuda 9.1
pytorch 0.4.0
cudnn is automatically installed by pytorch 0.4.0
system: centos 7.2.1511

The images below is captured when I run "test.py". The gpu memory gradully increased from 600MB -> 3800, and it will keep increasing until it occupied all of the gpu memory. 

<img width="742" alt="2018-07-23 11 04 30" src="https://user-images.githubusercontent.com/18735019/43055367-e17ef014-8e68-11e8-8916-c831a9f54882.png">

<img width="865" alt="2018-07-23 11 04 52" src="https://user-images.githubusercontent.com/18735019/43055408-f567fa44-8e68-11e8-8308-d954887dbc06.png">

**Comment by tamakoji** [comment]
hmm. I only have tested up to 4gpu. let me see if i can spare pocket money to test on p2.x8large

**Comment by weitaoatvison** [comment]
@tamakoji I tested the code on 4 gpu, it works, both on "test.py" and my code. So, I think the gpu memory error may caused by the number of gpu cards. Could you fix the bug? Or you can give me some suggestions, I will try to fix it.

**Comment by weitaoatvison** [comment]
@tamakoji I also found If I use sync-bn instead of standard bn, training will take more Gpu memory, is it right? In my test case, one net will cost about 5200MB on each card when I use standard bn, and it will cost 10000MB on each card when I use sync-bn, in this case, I used 4 cards in all in training.

**Comment by tamakoji** [comment]
@weitaoatvison the only thing I can doubt is [torch.cuda.comm](https://pytorch.org/docs/stable/_modules/torch/cuda/comm.html) module, syncbn has master and slaves. master will gather all sum(x),sum(x^2) (each num_feature size) from slaves and computes global mean&var. In order to gather tensors in different GPUs, torch.cuda.comm is used, and it also use NCCL as backend. There may be dependency problem of NCCL, likewise stated in https://discuss.pytorch.org/t/comm-broadcast-coalesced-not-returning/14961
for memory growth, yes since master gpu needs to gather all sum(x),sum(x^2) from slaves, it requires feature_size x num_gpu x 2(=sum,sumsq) memory. same thing happens in backward. and amount increase depends on feature_size.

**Comment by weitaoatvison** [comment]
@tamakoji Thanks！I will try to fix the bug. And if you have new update, please tell me. Thanks very much!

**Comment by d-li14** [comment]
Sorry, I have run `test.py` with cuda 9.1.85, pytorch 0.4.0, cudnn installed by pytorch on 2 GPUs, but the difference between outputs of built-in `nn.BatchNorm2d` and customized `modules.nn.BatchNorm2d` is too minor to be the same.

**Comment by tamakoji** [comment]
@d-li14 
SyncBN usually works on the task where you can't have much batchsize even with multi-gpu, such as Semantic Segmentation. For tasks like ILSVRC, the difference may be trivial. For your reference, see below as an sample ran on ILSVRC with ResNet50 using/not using SyncBN with minibatch size of 256(4gpux64) for first 39ep. the improvement could be said minor, but top5 is better on SyncBN.
![syncnsync_ilsvrc12_resnet50](https://user-images.githubusercontent.com/7410464/43532995-bd0355da-95ee-11e8-80c3-3fec70ed415d.png)
The aim that I made this implementation of SyncBN publicly available is to let people try on their task without much of the modification on the code but not to guarantee that it improves arbitrary tasks/datasets.

**Comment by d-li14** [comment]
@tamakoji 
Thanks for your detailed explanation, I knew that why we need synchronized bn. As you put it, it's designed for tasks in which minibatch size is not large enough on a single GPU such as segmentation. Instead, for ILSVRC, the batch size is large enough, even divided by the num of GPUs (for example, 256/4 = 56), so the performance gap between with and w/o sync bn is not so noticeable.

However, my original problem is merely about the `test.py` in your repo, I just want to conduct a simple test before applying it to my segmentation code and I wonder what the desired output of this test script is. As is known to us, in `test.py` batch size is set to be equal to num of GPUs. So as stated in the [Usage](https://github.com/tamakoji/pytorch-syncbn#usage) section of  [README.md](https://github.com/tamakoji/pytorch-syncbn/blob/master/README.md), it should be plausible that some differences exist between the behaviour of built-in `nn.BatchNorm2d` and customized `modules.nn.BatchNorm2d`, presuming the syncbn works well. But in my case, outputs of the two are almost the same unluckily.

**Comment by tamakoji** [comment]
@d-li14 If you have further comments, please create different thread since this issue topic was meant for memory issue with gpu >4.
The `test.py` was only intended for those cloned this repo and wanted to check the implementation of  `modules.nn.Batchnorm2d`  should be almost/exact same  to the original `nn.BatchNorm2d` if given *same* minibatch size. so `test.py` should give same answer and same stats learned.

**Comment by justttry** [comment]
@weitaoatvison do you fix the bug? I meet the same problem when I use 8 gpus. 4 gpus is OK.
